### PR TITLE
Issue #3888 - Huge file Resources behaviors

### PR DESCRIPTION
**Pull request by joakime**
Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

### CODE REVIEW

File: jetty-webapp/src/test/java/org/eclipse/jetty/webapp/HugeResourceTest.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        makeStaticFile(hugeStaticBase.resolve("test-1g.dat"), 1 * GB);
+        makeStaticFile(hugeStaticBase.resolve("test-4g.dat"), 4 * GB);
+        makeStaticFile(hugeStaticBase.resolve("test-10g.dat"), 10 * GB);
+
+        outputDir = MavenTestingUtils.getTargetTestingPath(HugeResourceTest.class.getSimpleName() + "-huge-static-outputdir");
+        FS.ensureEmpty(outputDir);
+    }
+
+    @AfterAll
+    public static void cleanupHugeStaticFiles()
```
Comment (sbordet) [inline jetty-webapp/src/test/java/org/eclipse/jetty/webapp/HugeResourceTest.java]: Remove "huge" everywhere, it's not really needed.

File: jetty-webapp/src/test/java/org/eclipse/jetty/webapp/HugeResourceTest.java [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+    @Test
+    public void testDownload_10G() throws IOException
+    {
+        download(server.getURI().resolve("/test-10g.dat"), 10 * GB);
+    }
+
+    private void download(URI destUri, long expectedSize) throws IOException
+    {
+        HttpURLConnection http = (HttpURLConnection)destUri.toURL().openConnection();
```
Comment (sbordet) [inline jetty-webapp/src/test/java/org/eclipse/jetty/webapp/HugeResourceTest.java]: Why, oh why??? Use `HttpClient`, so we also know if the client works well with these files.

File: jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java [hunk 3]
```diff
@@ -971,11 +971,19 @@ public static boolean isMappedBuffer(ByteBuffer buffer)
...
-        int len = (int)resource.length();
+        long len = resource.length();
         if (len < 0)
             throw new IllegalArgumentException("invalid resource: " + resource + " len=" + len);
 
-        ByteBuffer buffer = direct ? BufferUtil.allocateDirect(len) : BufferUtil.allocate(len);
+        if (len > Integer.MAX_VALUE)
+        {
+            // This method cannot handle resources of this size.
+            return null;
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java]: Are callers of this method prepared to receive `null`?
Before this change, this method was never returning `null`.
I would rather throw if it's an exceptional case.

Comment (sbordet) [reply jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java]: I also would like to see a test for this case (either if we return `null` or throw). It is already covered?

Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java]: Yes, there are only 4 callers for this `BufferUtil.toBuffer(Resource, boolean)` method.
Excluding test cases, there's only the 2 flavors of `.getDirectBuffer()` and `.getIndirectBuffer()` that use this method.
All 4 usages of `.getDirectBuffer()` and `.getIndirectBuffer()` handle null and expect it.

Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java]: Throwing causes an exception to be logged at WARN level every time.
This isn't necessary based on how `.getDirectBuffer()` and `.getIndirectBuffer()` are used.
If those are unavailable, (by returning null), then the `Resource.getReadableByteChannel()` or `Resource.getInputStream()` are used instead.

See HttpOutput.sendContent() for example usage:

https://github.com/eclipse/jetty.project/blob/4ebf4b7477acb42c73a069a37e29b5fd0738bfec/jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java#L987-L1020

File: jetty-webapp/src/test/java/org/eclipse/jetty/webapp/HugeResourceTest.java (hunk 1 above)
Comment (joakime) [reply jetty-webapp/src/test/java/org/eclipse/jetty/webapp/HugeResourceTest.java]: done.

File: jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java (hunk 3 above)
Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java]: After conversation with @gregw we added another safety check for buffers before the `BufferUtil.toBuffer(Resource, boolean)` usage in the CachedContentFactory.

File: jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java [hunk 4]
```diff
@@ -589,7 +601,7 @@ public ByteBuffer getDirectBuffer()
                     else
                         buffer = _mappedBuffer.get();
                 }
-                else
+                else if (_resource.length() < _maxCachedFileSize)
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java]: Need a comment here explaining this condition... and why it doesn't apply to mapped buffers above

File: jetty-server/src/test/java/org/eclipse/jetty/server/RequestTest.java [hunk 5]
```diff
@@ -639,6 +639,56 @@ public boolean check(HttpServletRequest request, HttpServletResponse response)
         assertThat(responses, startsWith("HTTP/1.1 200"));
     }
 
+    @Test
+    public void testContentLength_ExceedsMaxInteger() throws Exception
```
Comment (gregw) [inline jetty-server/src/test/java/org/eclipse/jetty/server/RequestTest.java]: not a fan of these _ in method names!

Comment (joakime) [reply jetty-server/src/test/java/org/eclipse/jetty/server/RequestTest.java]:
```
[jetty.project-9.4.x]$ $ find . -wholename "*/src/test/java/*.java" -exec grep "void test[A-Za-z0-9]*_[A-Za-z0-9]" {} \; | wc -l
934
```

We have 934 instances of this in our codebase.

File: jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java (hunk 4 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/CachedContentFactory.java]: Done. fixed in commit 602c44ec96966b0a40b3b3fae0e465a2a114ca1f
